sphinx/util/inspect.py

```diff
@@ -313,6 +315,98 @@ def is_builtin_class_method(obj: Any, attr_name: str) -> bool:
     return getattr(builtins, safe_getattr(cls, '__name__', '')) is cls
 
 
+def ast_unparse(node: ast.AST) -> str:
+    """Unparse an AST to string."""
+    if node is None:
+        return None
+    elif isinstance(node, ast.Attribute):
+        return '%s.%s' % (ast_unparse(node.value), node.attr)
+    elif isinstance(node, ast.Bytes):
+        return repr(node.s)
+    elif isinstance(node, ast.Call):
+        args = ([ast_unparse(e) for e in node.args] +
+                ['%s=%s' % (k.arg, ast_unparse(k.value)) for k in node.keywords])
+        return '%s(%s)' % (ast_unparse(node.func), ', '.join(args))
+    elif isinstance(node, ast.Dict):
+        keys = (ast_unparse(k) for k in node.keys)
+        values = (ast_unparse(v) for v in node.values)
+        items = (k + ': ' + v for k, v in zip(keys, values))
+        return '{' + ', '.join(items) + '}'
+    elif isinstance(node, ast.Ellipsis):
+        return '...'
+    elif sys.version_info > (3, 6) and isinstance(node, ast.Constant):
+        return repr(node.value)
+    elif isinstance(node, ast.Index):
+        return ast_unparse(node.value)
+    elif isinstance(node, ast.Lambda):
+        return '<function <lambda>>'  # TODO
+    elif isinstance(node, ast.List):
+        return '[' + ', '.join(ast_unparse(e) for e in node.elts) + ']'
+    elif isinstance(node, ast.Name):
+        return node.id
+    elif isinstance(node, ast.NameConstant):
+        return repr(node.value)
+    elif isinstance(node, ast.Num):
+        return repr(node.n)
+    elif isinstance(node, ast.Set):
+        return "{" + ', '.join(ast_unparse(e) for e in node.elts) + "}"
+    elif isinstance(node, ast.Str):
+        return repr(node.s)
+    elif isinstance(node, ast.Subscript):
+        return "%s[%s]" % (ast_unparse(node.value), ast_unparse(node.slice))
+    elif isinstance(node, ast.Tuple):
+        return ', '.join(ast_unparse(e) for e in node.elts)
+    else:
+        # TODO: f-string
+        raise NotImplementedError('Unable to render %s instance' % type(node).__name__)
+
+
+def signature_from_str(signature: str) -> inspect.Signature:
```

Comment:
It's possible there's a trick you can play here using `__text_signature__`, will see if I can construct it when I'm back at a shell

Reply:
Ah, I'm thinking of:
```
>>> inspect._signature_fromstr(inspect.Signature, lambda: None, "(a, b, c)")
<Signature (a, b, c)>
```
but unfortunately
```
>>> inspect._signature_fromstr(inspect.Signature, lambda: None, "(a, b, c: int)")
ValueError: Annotations are not currently supported
```
It would be great to fix this upstream in CPython

Reply:
As its docstring says, `_signature_fromstr()` is private. We have to discuss about exposure before fixing.
Anyway, it is not a scope of this PR. marked as resolved.

Reply:
I noticed it might be better to rename function name from "..._from_str" to "..._fromstr". But I don't mind either.

Reply:
> We have to discuss about exposure before fixing.

I think this patch would be valuable upstream whether sphinx uses it or not